File: week2/java/week2_6.java
public class week2_6 {
    public static int solution(int[][] triangle){
        int[][] dp = new int[triangle.length][triangle.length];

        dp[0][0] = triangle[0][0];
        for(int i=1; i<triangle.length; i++){
            dp[i][0] = dp[i-1][0] + triangle[i][0];
            dp[i][i] = dp[i-1][i-1] + triangle[i][i];
        }

        for(int i=2; i<triangle.length; i++){
            for(int j=1; j<i; j++){
                dp[i][j] = Math.max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j];
            }
        }

        int max = 0;
        for(int i=0; i<triangle.length; i++){
            max = Math.max(max, dp[triangle.length-1][i]);
        }
        return max;
    }
    public static void main(String[] args){

    }
}

File: week2/c++/week2_6.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int d[501][10000];

int solution(vector<vector<int>> triangle)
{

    // DP bottom-up
    for (int i = 0; i < triangle.size(); i++)
    {
        for (int j = 0; j < i + 1; j++)
        {

            if (i == 0)
            {
                d[i][j] += triangle[i][j];
            }

            else
            {
                if (j == 0)
                {
                    d[i][j] = d[i - 1][j] + triangle[i][j];
                }
                else if (j == i)
                {
                    d[i][j] = d[i - 1][j - 1] + triangle[i][j];
                }
                else
                {
                    d[i][j] = max(d[i - 1][j], d[i - 1][j - 1]) + triangle[i][j];
                }
            }
        }
    }

    // 맨 마지막 줄 d[i][j]의 최댓값
    int answer = d[triangle.size() - 1][0];
    for (int i = 1; i < triangle.size(); i++)
    {
        if (answer < d[triangle.size() - 1][i])
        {
            answer = d[triangle.size() - 1][i];
        }
    }

    return answer;
}